File: src/main/java/cc/hyperium/addons/AddonManifest.java
/*
 *     Copyright (C) 2018  Hyperium <https://hyperium.cc/>
 *
 *     This program is free software: you can redistribute it and/or modify
 *     it under the terms of the GNU Lesser General Public License as published
 *     by the Free Software Foundation, either version 3 of the License, or
 *     (at your option) any later version.
 *
 *     This program is distributed in the hope that it will be useful,
 *     but WITHOUT ANY WARRANTY; without even the implied warranty of
 *     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *     GNU Lesser General Public License for more details.
 *
 *     You should have received a copy of the GNU Lesser General Public License
 *     along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package cc.hyperium.addons;

import cc.hyperium.exceptions.HyperiumException;
import com.google.common.io.Files;
import com.google.gson.JsonObject;
import com.google.gson.JsonParser;
import org.apache.commons.io.IOUtils;

import java.io.File;
import java.io.FileOutputStream;
import java.io.InputStream;
import java.io.OutputStream;
import java.nio.charset.Charset;
import java.util.jar.JarFile;
import java.util.zip.ZipEntry;

public class AddonManifest {

    /**
     * Manifest json
     */
    private JsonObject json;

    /**
     * @param jar Addon jar
     * @throws HyperiumException if it fails to read description
     */
    public AddonManifest(JarFile jar) throws HyperiumException {

        try {
            ZipEntry entry = jar.getEntry("addon.json");
            JsonParser parser = new JsonParser();

            File jsonFile = File.createTempFile("json", "tmp");
            jsonFile.deleteOnExit();

            InputStream jarInputStream = jar.getInputStream(entry);
            OutputStream os = new FileOutputStream(jsonFile);
            IOUtils.copy(jarInputStream, os);

            JsonObject json = parser.parse(Files.toString(jsonFile, Charset.defaultCharset())).getAsJsonObject();
            this.json = json;
            if (!json.has("version") && !json.has("name") && !json.has("main")) {
                throw new HyperiumException("Invalid addon jar (addon.json does not exist or invalid)");
            }
        } catch (Exception e) {
            throw new HyperiumException("Invalid addon jar (addon.json does not exist or invalid)");
        }
    }

    /**
     * @return addon name
     */
    public String getName() {
        return json.get("name").getAsString();
    }

    /**
     * @return addon version
     */
    public String getVersion() {
        return json.get("version").getAsString();
    }

    /**
     * @return main-class
     */
    public String getMain() {
        return json.get("main").getAsString();
    }

    /**
     * @return addon manifest json in string
     */
    @Override
    public String toString() {
        return json.toString();
    }
}

File: src/main/kotlin/cc/hyperium/internal/addons/AddonManifest.kt
/*
 *       Copyright (C) 2018-present Hyperium <https://hyperium.cc/>
 *
 *       This program is free software: you can redistribute it and/or modify
 *       it under the terms of the GNU Lesser General Public License as published
 *       by the Free Software Foundation, either version 3 of the License, or
 *       (at your option) any later version.
 *
 *       This program is distributed in the hope that it will be useful,
 *       but WITHOUT ANY WARRANTY; without even the implied warranty of
 *       MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *       GNU Lesser General Public License for more details.
 *
 *       You should have received a copy of the GNU Lesser General Public License
 *       along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package cc.hyperium.internal.addons

/**
 * Instance created on the classloader net.minecraft.launchwrapper.LaunchClassLoader
 *
 * @since 1.0
 * @author Kevin Brewster
 */
class AddonManifest {

    /**
     * The Identifier the addon goes by.
     * <i>This should be unique</i>
     * @return addon name
     */
    var name: String? = null

    /**
     * Version of the addon
     * <i>Preferred versioning number pattern is <b>MAJOR.MINOR.PATCH</b></i>
     * MAJOR - version when you make incompatible API changes.
     * MINOR - version when you add functionality in a backwards-compatible manner.
     * PATCH - version when you make backwards-compatible bug fixes.
     *
     * @return addon version
     */
    var version: String? = null

    /**
     * Description of an Addon
     *
     * @return desc
     */

    val desc: String? = null

    /**
     * Main Class of an addon
     * e.g.com.example.ExampleAddon
     *
     * @return main-class
     */
    val mainClass: String? = null

    /**
     * If the addon includes and Mixins you
     * can add the config(s) here which will
     * automatically at them to the environment
     *
     * @return mixin configs
     */
    val mixinConfigs: List<String>? = null

    /**
     * If the addon includes a <i>tweaker</i>
     * you can specify it here and it will
     * add it to the launchwrapper.
     * <b>IT MUST IMPLEMENT {@link net.minecraft.launchwrapper.ITweaker}!</b>
     *
     * @return tweaker-class
     */
    val tweakerClass: String? = null

    /**
     * An array containing all the names of
     * the dependencies. The dependencies
     * should be loaded before this addon.
     *
     * @return the names of the dependencies
     */
    val dependencies: List<String> = arrayListOf()

    /**
     * If the addon includes a {@link net.minecraft.launchwrapper.IClassTransformer}
     * you can specify it here.
     */
    val transformerClass: String? = null

    /**
     * Class where the config for the addon is
     */
    val overlay: String? = null

    /**
     * The addon developer
     */
    val author: String? = null

    /**
     * Must match up to AddonMinecraftBootstrap#VERSION_CODE
     */
    val versionCode: String? = null
}
